File: sdk/javascript/src/merge.js
/**
 * FoodBlock Merge — conflict resolution for forked update chains.
 * Merges turn chains into DAGs.
 */

const { create } = require('./block')

/**
 * Detect whether two block hashes represent a fork (conflict) in an
 * update chain by walking both chains backward to find a common ancestor.
 *
 * @param {string} hashA - First chain head hash
 * @param {string} hashB - Second chain head hash
 * @param {function} resolve - async (hash) => block | null
 * @returns {object} { isConflict, commonAncestor, chainA, chainB }
 */
async function detectConflict(hashA, hashB, resolve) {
  if (hashA === hashB) {
    return { isConflict: false, commonAncestor: hashA, chainA: [], chainB: [] }
  }

  const chainA = []
  const chainB = []
  const visitedA = new Set()
  const visitedB = new Set()

  // Walk chain A
  let current = hashA
  while (current) {
    visitedA.add(current)
    const block = await resolve(current)
    if (!block) break
    chainA.push(block)
    current = block.refs && block.refs.updates
      ? (Array.isArray(block.refs.updates) ? block.refs.updates[0] : block.refs.updates)
      : null
  }

  // Walk chain B, looking for intersection with A
  let commonAncestor = null
  current = hashB
  while (current) {
    if (visitedA.has(current)) {
      commonAncestor = current
      break
    }
    visitedB.add(current)
    const block = await resolve(current)
    if (!block) break
    chainB.push(block)
    current = block.refs && block.refs.updates
      ? (Array.isArray(block.refs.updates) ? block.refs.updates[0] : block.refs.updates)
      : null
  }

  return {
    isConflict: commonAncestor !== null,
    commonAncestor,
    chainA,
    chainB
  }
}

/**
 * Create a merge block that resolves a fork between two chain heads.
 *
 * @param {string} hashA - First fork head hash
 * @param {string} hashB - Second fork head hash
 * @param {function} resolve - async (hash) => block | null
 * @param {object} [opts] - { state, strategy: 'a_wins' | 'b_wins' | 'manual' }
 * @returns {object} The merge FoodBlock
 */
async function merge(hashA, hashB, resolve, opts = {}) {
  const strategy = opts.strategy || 'manual'

  let mergedState
  if (strategy === 'manual') {
    if (!opts.state) {
      throw new Error('FoodBlock: manual merge requires opts.state')
    }
    mergedState = opts.state
  } else if (strategy === 'a_wins') {
    const blockA = await resolve(hashA)
    if (!blockA) throw new Error('FoodBlock: could not resolve hashA')
    mergedState = blockA.state
  } else if (strategy === 'b_wins') {
    const blockB = await resolve(hashB)
    if (!blockB) throw new Error('FoodBlock: could not resolve hashB')
    mergedState = blockB.state
  } else {
    throw new Error('FoodBlock: unknown merge strategy: ' + strategy)
  }

  return create('observe.merge', {
    strategy,
    ...mergedState
  }, {
    merges: [hashA, hashB]
  })
}

/**
 * Attempt an automatic merge using vocabulary-defined per-field strategies.
 *
 * @param {string} hashA - First fork head hash
 * @param {string} hashB - Second fork head hash
 * @param {function} resolve - async (hash) => block | null
 * @param {object} [vocabulary] - A vocabulary block with merge strategies per field
 * @returns {object} The merge FoodBlock
 */
async function autoMerge(hashA, hashB, resolve, vocabulary) {
  const blockA = await resolve(hashA)
  const blockB = await resolve(hashB)

  if (!blockA) throw new Error('FoodBlock: could not resolve hashA')
  if (!blockB) throw new Error('FoodBlock: could not resolve hashB')

  const stateA = blockA.state || {}
  const stateB = blockB.state || {}

  // Gather all keys from both states
  const allKeys = new Set([...Object.keys(stateA), ...Object.keys(stateB)])
  const mergedState = {}
  const fields = vocabulary && (vocabulary.state || vocabulary).fields

  for (const key of allKeys) {
    const valA = stateA[key]
    const valB = stateB[key]

    // If values are the same, no conflict
    if (JSON.stringify(valA) === JSON.stringify(valB)) {
      mergedState[key] = valA !== undefined ? valA : valB
      continue
    }

    // If only one side has the value, take it
    if (valA === undefined) { mergedState[key] = valB; continue }
    if (valB === undefined) { mergedState[key] = valA; continue }

    // Values differ — use vocabulary strategy if available
    const fieldDef = fields && fields[key]
    const mergeStrategy = fieldDef && fieldDef.merge

    if (!mergeStrategy || mergeStrategy === 'conflict') {
      throw new Error('FoodBlock: auto-merge conflict on field "' + key + '" — manual resolution required')
    }

    if (mergeStrategy === 'last_writer_wins' || mergeStrategy === 'lww') {
      // Prefer B (convention: later writer)
      mergedState[key] = valB
    } else if (mergeStrategy === 'max') {
      mergedState[key] = typeof valA === 'number' && typeof valB === 'number'
        ? Math.max(valA, valB) : valB
    } else if (mergeStrategy === 'min') {
      mergedState[key] = typeof valA === 'number' && typeof valB === 'number'
        ? Math.min(valA, valB) : valB
    } else if (mergeStrategy === 'union') {
      const arrA = Array.isArray(valA) ? valA : [valA]
      const arrB = Array.isArray(valB) ? valB : [valB]
      mergedState[key] = [...new Set([...arrA, ...arrB])]
    } else {
      throw new Error('FoodBlock: unknown merge strategy "' + mergeStrategy + '" for field "' + key + '"')
    }
  }

  return create('observe.merge', {
    strategy: 'auto',
    ...mergedState
  }, {
    merges: [hashA, hashB]
  })
}

module.exports = { detectConflict, merge, autoMerge }

File: sdk/go/merge.go
package foodblock

import (
	"encoding/json"
	"errors"
)

// ConflictResult holds the result of a conflict detection.
type ConflictResult struct {
	IsConflict     bool
	CommonAncestor string
	ChainA         []Block
	ChainB         []Block
}

// DetectConflict detects whether two hashes represent a fork in an update chain.
func DetectConflict(hashA, hashB string, resolve func(string) *Block) ConflictResult {
	if hashA == hashB {
		return ConflictResult{IsConflict: false, CommonAncestor: hashA}
	}

	var chainA []Block
	visitedA := make(map[string]bool)

	// Walk chain A
	current := hashA
	for current != "" {
		visitedA[current] = true
		block := resolve(current)
		if block == nil {
			break
		}
		chainA = append(chainA, *block)
		if updates, ok := block.Refs["updates"]; ok {
			if s, ok := updates.(string); ok {
				current = s
			} else {
				current = ""
			}
		} else {
			current = ""
		}
	}

	// Walk chain B, looking for intersection with A
	var chainB []Block
	commonAncestor := ""
	current = hashB
	for current != "" {
		if visitedA[current] {
			commonAncestor = current
			break
		}
		block := resolve(current)
		if block == nil {
			break
		}
		chainB = append(chainB, *block)
		if updates, ok := block.Refs["updates"]; ok {
			if s, ok := updates.(string); ok {
				current = s
			} else {
				current = ""
			}
		} else {
			current = ""
		}
	}

	return ConflictResult{
		IsConflict:     commonAncestor != "",
		CommonAncestor: commonAncestor,
		ChainA:         chainA,
		ChainB:         chainB,
	}
}

// Merge creates a merge block that resolves a fork between two chain heads.
// strategy can be "manual", "a_wins", or "b_wins".
func Merge(hashA, hashB string, resolve func(string) *Block, strategy string, manualState map[string]interface{}) (Block, error) {
	if strategy == "" {
		strategy = "manual"
	}

	var mergedState map[string]interface{}

	switch strategy {
	case "manual":
		if manualState == nil {
			return Block{}, errors.New("FoodBlock: manual merge requires state")
		}
		mergedState = manualState
	case "a_wins":
		blockA := resolve(hashA)
		if blockA == nil {
			return Block{}, errors.New("FoodBlock: could not resolve hashA")
		}
		mergedState = blockA.State
	case "b_wins":
		blockB := resolve(hashB)
		if blockB == nil {
			return Block{}, errors.New("FoodBlock: could not resolve hashB")
		}
		mergedState = blockB.State
	default:
		return Block{}, errors.New("FoodBlock: unknown merge strategy: " + strategy)
	}

	state := map[string]interface{}{"strategy": strategy}
	for k, v := range mergedState {
		state[k] = v
	}

	return Create("observe.merge", state, map[string]interface{}{
		"merges": []interface{}{hashA, hashB},
	}), nil
}

// AutoMerge attempts automatic merge using per-field strategies from a vocabulary.
func AutoMerge(hashA, hashB string, resolve func(string) *Block, fieldStrategies map[string]string) (Block, error) {
	blockA := resolve(hashA)
	blockB := resolve(hashB)
	if blockA == nil {
		return Block{}, errors.New("FoodBlock: could not resolve hashA")
	}
	if blockB == nil {
		return Block{}, errors.New("FoodBlock: could not resolve hashB")
	}

	stateA := blockA.State
	stateB := blockB.State
	if stateA == nil {
		stateA = map[string]interface{}{}
	}
	if stateB == nil {
		stateB = map[string]interface{}{}
	}

	allKeys := make(map[string]bool)
	for k := range stateA {
		allKeys[k] = true
	}
	for k := range stateB {
		allKeys[k] = true
	}

	mergedState := map[string]interface{}{}
	for key := range allKeys {
		valA := stateA[key]
		valB := stateB[key]

		// If values are the same, no conflict
		jsonA, _ := json.Marshal(valA)
		jsonB, _ := json.Marshal(valB)
		if string(jsonA) == string(jsonB) {
			if valA != nil {
				mergedState[key] = valA
			} else {
				mergedState[key] = valB
			}
			continue
		}

		if valA == nil {
			mergedState[key] = valB
			continue
		}
		if valB == nil {
			mergedState[key] = valA
			continue
		}

		// Values differ — use strategy
		strategy := ""
		if fieldStrategies != nil {
			strategy = fieldStrategies[key]
		}

		switch strategy {
		case "last_writer_wins", "lww":
			mergedState[key] = valB
		case "max":
			fA, okA := toFloat64(valA)
			fB, okB := toFloat64(valB)
			if okA && okB {
				if fA > fB {
					mergedState[key] = valA
				} else {
					mergedState[key] = valB
				}
			} else {
				mergedState[key] = valB
			}
		case "min":
			fA, okA := toFloat64(valA)
			fB, okB := toFloat64(valB)
			if okA && okB {
				if fA < fB {
					mergedState[key] = valA
				} else {
					mergedState[key] = valB
				}
			} else {
				mergedState[key] = valB
			}
		default:
			return Block{}, errors.New("FoodBlock: auto-merge conflict on field \"" + key + "\" — manual resolution required")
		}
	}

	state := map[string]interface{}{"strategy": "auto"}
	for k, v := range mergedState {
		state[k] = v
	}

	return Create("observe.merge", state, map[string]interface{}{
		"merges": []interface{}{hashA, hashB},
	}), nil
}

func toFloat64(v interface{}) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	default:
		return 0, false
	}
}
